man/cafe-keyboard-properties.1
.\" Copyright (C) 2007 Sven Arvidsson <sa@whiz.se>
.\" Copyright (C) 2014 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.TH cafe-keyboard-properties 1 "2014\-05\-02" "CAFE"
.SH NAME
cafe-keyboard-properties \- manage keyboard behaviour in CAFE
.SH SYNOPSIS
.B cafe-keyboard-properties
.RI [ OPTIONS... ]
.SH DESCRIPTION
This executable is part of the package 'cafe\-control\-center': The CAFE Control Center.
.PP
Keyboard Preferences manages all settings related to the keyboard in a
CAFE environment. This includes key repeat delay, cursor blinking,
layouts and layout options. You can also configure a typing break
which locks the screen and enforces a break from work.
.SH OPTIONS
.TP
.B \-\-apply
Just apply settings and quit (compatibility only; now handled by daemon)
.TP
.B \-\-init\-session\-settings
Just apply settings and quit (compatibility only; now handled by daemon)
.TP
.B \-\-get\-legacy
Retrieve and store legacy settings
.TP
.B \-\-typing-break
Start the page with the typing break settings showing
.P
This program also accepts the standard CAFE and CTK options.
.SH AUTHORS
This manual page was writtenby Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
for the Debian project (but may be used by others).
.SH SEE ALSO
.BR "cafe-control-center" (1),
.BR "ctk-options" (7)


NEWS
### cafe-control-center 1.24.1

  * Translations update
  * cafe-theme-info: Use of memory after it is freed
  * build: silent build warnings for distcheck
  * use default CAFE_COMPILE_WARNINGS level
  * Port libslab_handle_g_error to the built-in GLib logging framework
  * cafe-about-me: Fix memory leak
  * appearance: Fix memory leak
  * Do not collect the translation for Icon

### cafe-control-center 1.24.0

  * Translations update
  * cafe-theme-info: fix memory leak
  * cafe-wp-info: fix memory leak
  * glib-genmarshal: Using --header and --body at the same time is deprecated
  * build: show compiler/flags info on configure summary
  * build: do not override WARN_CFLAGS
  * build: use warning flags
  * time-admin: make portable backward file path
  * appearance: backgrounds dir should be portable
  * appearance: remove I/O error on window close
  * autoconf: AC_PROG_CC was called before AX_CHECK_ENABLE_DEBUG
  * Fix build using gcc 10 -fno-common flag

### cafe-control-center 1.23.3

  * Translations update
  * Update copyright to 2020
  * about-me: Fix UI validation
  * mouse: Fix UI validation
  * time-admin: Add license info in time-admin.ui file
  * time-admin: Fix UI validation
  * at-properties: Fix UI validation
  * Search system themes in system data dirs
  * display: Fix UI validation
  * display: Enable help button
  * windows: Add window-properties.ui
  * avoid redundant redeclarations
  * keybindings: Fix schema creation in makefile
  * appearance: Fix desktop file creation
  * shell: Fix typo in keyword name
  * migrate from intltool to gettext
  * default-applications: Load icons as cairo surfaces
  * icons: Force uniform size on icons (#505)
  * remove warning: "__sun" is not defined, evaluates to 0
  * fv: thumbnailer: add font/collection to the supported mimetypes
  * font-viewer: Add font/ttf and font/otf mime types
  * time-admin: fix memory leak
  * time-admin: Don't share «void QuitApp(TimeAdmin *ta)» method
  * default-applications: Remove messenger options from ui file
  * time-admin: Remove unreachable code
  * time-admin: Simplify date & time displaying
  * time-admin: Save button cannot be enabled if NTP sync is active
  * time-admin: Display current time in time-admin.ui on startup
  * time-admin: Rename adjustament to adjustment in time-admin.ui
  * time-admin: A11y time-admin.ui
  * time-admin: Add time-admin.ui
  * default-applications: Fix Word Processor top margin

### cafe-control-center 1.23.2

  * Translations update
  * travis: update cafe-desktop and cafe-settings-daemon required versions
  * mouse: Add support for Acceleration Profiles
  * appearance: fix memory leak
  * appearance: add new font for user
  * Stop propagate error
  * at-properties: migrate from dbus-glib to gdbus

### cafe-control-center 1.23.1

  * Translations update
  * fix build error when not installed gla11y
  * remove lightdm-settings support
  * launch lightdm-settings if installed
  * use current running DM preferences program
  * support lightdm-ctk-greeter-settings as login preferences program
  * replace Tab with space
  * drop libXxf86 dependency
  * Use python3 for gla11y
  * Add scroll tabs with mouse wheel in severals capplets
  * libslab: bookmark-agent.c avoid deprecated g_type_class_add_private
  * font-viewer: avoid deprecated 'g_type_class_add_private'
  * Remove compilation warnings
  * time-admin: Don't set window title for message-box dialog
  * time-admin: Don't set font size/color for regular controls
  * use theme's icon preferences-system-time
  * time-admin: Fix license notice in time-zone.h|c, time-map.h|c
  * capplets: Remove the macro CAFELOCALEDIR where it's not required
  * capplets: Use capplet_init
  * time-admin: shortcuts, images & shared code for buttons
  * common: Remove capplet-stock-icons
  * time-admin: Remove po-timezones
  * Remove trailing whitespaces

### cafe-control-center 1.23.0

  * Translations update
  * bump require of m-s-d and m-d
  * Add orca to Visual combo box if available
  * Add support for checking .ui accessibility with gla11y
  * Do not reload NTP and modify the button arrangement when starting the service
  * Authentication to add modification time
  * Use default color to set label and remove the maximum button
  * use Makefile pkgdatadir
  * g_date_time_unref datetime
  * Display the correct time when selecting the time zone
  * Memory leak: g_error_free(error);
  * Modify cafe-time-admin.desktop.in file format
  * enable to build time-admin for cafe-control-center
  * add time-admin source code
  * add timezone translation files for time-admin
  * add image resources for time-admin applet
  * Populate initial messenger list with irc scheme handler apps
  * Set default icq/irc/ircs/sip/xmpp scheme handler
  * Add instant messenger keybinding into cafe-keybinding-properties
  * Change the icon for instant messaging application
  * default-applications: Add a widget for a default messenger application
  * appearance.ui (cursor size scale)
  * libslab: avoid deprecated g_type_class_add_private
  * Create FUNDING.yml
  * typing-break: avoid deprecated 'g_type_class_add_private'
  * file-transfer-dialog.c: avoid deprecated 'g_type_class_add_private'
  * Fix the icon sizes in Preferred Applications selector
  * capplets: Use gresources for UI files
  * [ci] Add cppcheck html report
  * Fix grammar mistake
  * Migrate from dbus-glib to GDBus
  * Use system data dirs to locate key bindings
  * control-center: Fix memory leak
  * Use CtkSeparatorMenuItem in Apparence preview
  * Remove dead store detected by clang
  * default-applications: Improve UI accessibility
  * Use keyboard device icon like in other hardware devices
  * Use icons provided by themes and remove not used icons
  * help: update copyright
  * [ci] Enable Clang Static Analyzer
  * add installation instructions

### cafe-control-center 1.22.0

  * Translations update
  * Initialize travis support

capplets/keyboard/cafe-keyboard-properties.c
/* -*- mode: c; style: linux -*- */

/* keyboard-properties.c
 * Copyright (C) 2000-2001 Ximian, Inc.
 * Copyright (C) 2001 Jonathan Blandford
 *
 * Written by: Bradford Hovinen <hovinen@ximian.com>
 *             Rachel Hestilow <hestilow@ximian.com>
 *	       Jonathan Blandford <jrb@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdlib.h>
#include <gio/gio.h>

#include "capplet-util.h"
#include "activate-settings-daemon.h"

#include "cafe-keyboard-properties-a11y.h"
#include "cafe-keyboard-properties-xkb.h"

#define KEYBOARD_SCHEMA "org.cafe.peripherals-keyboard"
#define INTERFACE_SCHEMA "org.cafe.interface"
#define TYPING_BREAK_SCHEMA "org.cafe.typing-break"

enum {
	RESPONSE_APPLY = 1,
	RESPONSE_CLOSE
};

static GSettings * keyboard_settings = NULL;
static GSettings * interface_settings = NULL;
static GSettings * typing_break_settings = NULL;

static CtkBuilder *
create_dialog (void)
{
	CtkBuilder *dialog;
	CtkSizeGroup *size_group;
	CtkWidget *image;
	GError *error = NULL;

	dialog = ctk_builder_new ();
	if (ctk_builder_add_from_resource (dialog, "/org/cafe/mcc/keyboard/cafe-keyboard-properties-dialog.ui", &error) == 0) {
		g_warning ("Could not load UI: %s", error->message);
		g_error_free (error);
		g_object_unref (dialog);
		return NULL;
	}

	size_group = ctk_size_group_new (CTK_SIZE_GROUP_HORIZONTAL);
	ctk_size_group_add_widget (size_group, WID ("repeat_slow_label"));
	ctk_size_group_add_widget (size_group, WID ("delay_short_label"));
	ctk_size_group_add_widget (size_group, WID ("blink_slow_label"));
	g_object_unref (G_OBJECT (size_group));

	size_group = ctk_size_group_new (CTK_SIZE_GROUP_HORIZONTAL);
	ctk_size_group_add_widget (size_group, WID ("repeat_fast_label"));
	ctk_size_group_add_widget (size_group, WID ("delay_long_label"));
	ctk_size_group_add_widget (size_group, WID ("blink_fast_label"));
	g_object_unref (G_OBJECT (size_group));

	size_group = ctk_size_group_new (CTK_SIZE_GROUP_HORIZONTAL);
	ctk_size_group_add_widget (size_group, WID ("repeat_delay_scale"));
	ctk_size_group_add_widget (size_group, WID ("repeat_speed_scale"));
	ctk_size_group_add_widget (size_group, WID ("cursor_blink_time_scale"));
	g_object_unref (G_OBJECT (size_group));

	image = ctk_image_new_from_icon_name ("list-add", CTK_ICON_SIZE_BUTTON);
	ctk_button_set_image (CTK_BUTTON (WID ("xkb_layouts_add")), image);

	image = ctk_image_new_from_icon_name ("view-refresh", CTK_ICON_SIZE_BUTTON);
	ctk_button_set_image (CTK_BUTTON (WID ("xkb_reset_to_defaults")), image);

	return dialog;
}

static void
dialog_response (CtkWidget * widget,
		 gint response_id, guint data)
{
	if (response_id == CTK_RESPONSE_HELP)
		capplet_help (CTK_WINDOW (widget), "goscustperiph-2");
	else
		ctk_main_quit ();
}

static void
setup_dialog (CtkBuilder * dialog)
{
	gchar *monitor;

	g_settings_bind (keyboard_settings,
					 "repeat",
					 WID ("repeat_toggle"),
					 "active",
					 G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (keyboard_settings,
					 "repeat",
					 WID ("repeat_table"),
					 "sensitive",
					 G_SETTINGS_BIND_DEFAULT);

	g_settings_bind (keyboard_settings,
					 "delay",
					 ctk_range_get_adjustment (CTK_RANGE (WID ("repeat_delay_scale"))),
					 "value",
					 G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (keyboard_settings,
					 "rate",
					 ctk_range_get_adjustment (CTK_RANGE (WID ("repeat_speed_scale"))),
					 "value",
					 G_SETTINGS_BIND_DEFAULT);

	g_settings_bind (interface_settings,
					 "cursor-blink",
					 WID ("cursor_toggle"),
					 "active",
					 G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (interface_settings,
					 "cursor-blink",
					 WID ("cursor_hbox"),
					 "sensitive",
					 G_SETTINGS_BIND_DEFAULT);
	g_settings_bind (interface_settings,
					 "cursor-blink-time",
					 ctk_range_get_adjustment (CTK_RANGE (WID ("cursor_blink_time_scale"))),
					 "value",
					 G_SETTINGS_BIND_DEFAULT);

	/* Ergonomics */
	monitor = g_find_program_in_path ("cafe-typing-monitor");
	if (monitor != NULL) {
		g_free (monitor);

		g_settings_bind (typing_break_settings,
						 "enabled",
						 WID ("break_enabled_toggle"),
						 "active",
						 G_SETTINGS_BIND_DEFAULT);
		g_settings_bind (typing_break_settings,
						 "enabled",
						 WID ("break_details_table"),
						 "sensitive",
						 G_SETTINGS_BIND_DEFAULT);
		g_settings_bind (typing_break_settings,
						 "type-time",
						 WID ("break_enabled_spin"),
						 "value",
						 G_SETTINGS_BIND_DEFAULT);
		g_settings_bind (typing_break_settings,
						 "break-time",
						 WID ("break_interval_spin"),
						 "value",
						 G_SETTINGS_BIND_DEFAULT);
		g_settings_bind (typing_break_settings,
						 "allow-postpone",
						 WID ("break_postponement_toggle"),
						 "active",
						 G_SETTINGS_BIND_DEFAULT);

	} else {
		/* don't show the typing break tab if the daemon is not available */
		CtkNotebook *nb = CTK_NOTEBOOK (WID ("keyboard_notebook"));
		gint tb_page = ctk_notebook_page_num (nb, WID ("break_enabled_toggle"));
		ctk_notebook_remove_page (nb, tb_page);
	}

	g_signal_connect (WID ("keyboard_dialog"), "response",
			  (GCallback) dialog_response, NULL);

	setup_xkb_tabs (dialog);
	setup_a11y_tabs (dialog);
}

int
main (int argc, char **argv)
{
	CtkBuilder *dialog;
	GOptionContext *context;

	static gboolean apply_only = FALSE;
	static gboolean switch_to_typing_break_page = FALSE;
	static gboolean switch_to_a11y_page = FALSE;

	static GOptionEntry cap_options[] = {
		{"apply", 0, 0, G_OPTION_ARG_NONE, &apply_only,
		 N_
		 ("Just apply settings and quit (compatibility only; now handled by daemon)"),
		 NULL},
		{"init-session-settings", 0, 0, G_OPTION_ARG_NONE,
		 &apply_only,
		 N_
		 ("Just apply settings and quit (compatibility only; now handled by daemon)"),
		 NULL},
		{"typing-break", 0, 0, G_OPTION_ARG_NONE,
		 &switch_to_typing_break_page,
		 N_
		 ("Start the page with the typing break settings showing"),
		 NULL},
		{"a11y", 0, 0, G_OPTION_ARG_NONE,
		 &switch_to_a11y_page,
		 N_
		 ("Start the page with the accessibility settings showing"),
		 NULL},
		{NULL}
	};


	context = g_option_context_new (_("- CAFE Keyboard Preferences"));
	g_option_context_add_main_entries (context, cap_options,
					   GETTEXT_PACKAGE);

	capplet_init (context, &argc, &argv);

	activate_settings_daemon ();

	keyboard_settings = g_settings_new (KEYBOARD_SCHEMA);
	interface_settings = g_settings_new (INTERFACE_SCHEMA);
	typing_break_settings = g_settings_new (TYPING_BREAK_SCHEMA);

	dialog = create_dialog ();
	if (!dialog) /* Warning was already printed to console */
		exit (EXIT_FAILURE);

	setup_dialog (dialog);

        CtkNotebook* nb = CTK_NOTEBOOK (WID ("keyboard_notebook"));
        ctk_widget_add_events (CTK_WIDGET (nb), CDK_SCROLL_MASK);
        g_signal_connect (CTK_WIDGET (nb), "scroll-event",
                          G_CALLBACK (capplet_dialog_page_scroll_event_cb),
                          CTK_WINDOW (WID ("keyboard_dialog")));

	if (switch_to_typing_break_page) {
		ctk_notebook_set_current_page (nb, 4);
	}
	else if (switch_to_a11y_page) {
		ctk_notebook_set_current_page (nb, 2);

	}

	capplet_set_icon (WID ("keyboard_dialog"),
			  "input-keyboard");
	ctk_widget_show (WID ("keyboard_dialog"));
	ctk_main ();

	finalize_a11y_tabs ();
	g_object_unref (keyboard_settings);
	g_object_unref (interface_settings);
	g_object_unref (typing_break_settings);

	return 0;
}
